strawberryfields/api_client.py

```diff
@@ -0,0 +1,475 @@
+        }
+
+        # Try getting everything first from environment variables
+        configuration.update(self.get_configuration_from_environment())
+
+        # Override any values that are explicitly passed when initializing client
+        configuration.update(kwargs)
+
+        if configuration["hostname"] is None:
+            raise ValueError("hostname parameter is missing")
+
+        if configuration["hostname"] not in self.ALLOWED_HOSTNAMES:
+            raise ValueError("hostname parameter not in allowed list")
+
+        self.USE_SSL = configuration["use_ssl"]
+        if not self.USE_SSL:
+            warnings.warn("Connecting insecurely to API server", UserWarning)
+
+        self.HOSTNAME = configuration["hostname"]
+        self.BASE_URL = f"{'https' if self.USE_SSL else 'http'}://{self.HOSTNAME}"
+        self.AUTHENTICATION_TOKEN = configuration["authentication_token"]
+        self.HEADERS = {}
+
+        # TODO: warn if no authentication token
+
+    def get_configuration_from_environment(self):
+        """
+        Retrieve configuration from environment variables. The variables are defined as follows:
+        - SF_API_USE_SSL: True or False
+        - SF_API_HOSTNAME: The hostname of the server to connect to
+        - SF_API_AUTHENTICATION_TOKEN: The authentication token to use when connecting to the API
+        """
+        configuration = {
+            "authentication_token": os.environ.get(self.ENV_AUTHENTICATION_TOKEN_KEY),
+            "hostname": os.environ.get(self.ENV_API_HOSTNAME_KEY),
+            "use_ssl": os.environ.get(self.ENV_USE_SSL_KEY),
+        }
+
+        return {key: value for key, value in configuration.items() if value is not None}
+
+    def load_configuration_from_file(self):
+        """
+        Loads username, password, and/or authentication token from a config
+        file.
+        """
+        raise NotImplementedError()
+
+    def authenticate(self, username, password):
+        """
+        Retrieve an authentication token from the server via username
+        and password authentication.
+        """
+        raise NotImplementedError()
+
+    def set_authorization_header(self, authentication_token):
+        """
+        Adds the authorization header to the headers dictionary to be included
+        with all API requests.
+        """
+        self.headers["Authorization"] = authentication_token
+
+    def join_path(self, path):
+        """
+        Joins a base url with an additional path (e.g. a resource name and ID)
+        """
+        return join_path(self.BASE_URL, path)
+
+    def get(self, path):
+        """
+        Sends a GET request to the provided path. Returns a response object.
+        """
+        try:
+            response = requests.get(url=self.join_path(path), headers=self.HEADERS)
+        except requests.exceptions.ConnectionError as e:
+            response = None
+            warnings.warn(f"Could not connect to server ({e})")
```

Comment:
Since this isn't a core failure of the library (and rather, potentially a temporary error), I figured it might make sense to only show a warning and not a full blown exception. In theory I could do nothing here and let the exception surface. I have a Trello card for improving messaging and error handling altogether, in which I will include a better implementation of this, as well as all other client-server errors.